### ISSUE TITLE
Plugin support

### ISSUE THREAD

**Issue by delvedor** [fastify/fastify#39]
Before starting to implement the template (#30) and authentication (#10) support we must figure out how expose some APIs to build plugins for fastify without changing the core.

This issue is a placeholder for discussion.
cc @mcollina @davidmarkclements @lucamaraschi @marcbachmann

**Comment by marcbachmann** [comment]
Was there a discussion about that topic already?

> we must figure out how expose some APIs to build plugins for fastify without changing the core

I don't see any reason why not to do api changes changes.
Most times it's possible to do it backwards compatible.

First, what's the responsibility of plugins? There might be several kinds of plugins.
I guess we can differentiate between services (e.g. template engine) and request lifecycle plugins (e.g. authorization) for now.

IMO the services can mostly live outside of this module. As @jsumners suggested in https://github.com/fastify/fastify/issues/30, you can just instantiate sth by yourself and inject it into some other module. Maybe we can just offer some registry and some async way to register them before the server start.

Here are some notes about how my projects are currently set up.
That's an example image api endpoint (maybe not enough context here 😁):
```js
module.exports = function (feature, server, done) {
  // Parse Configuration
  const imageConfig = server.config.get('images')
  const featureConfig = require('./process_images_config')(imageConfig)

  // Initialize api (service), expose it for other apis/controllers and for tests
  const imagesApi = require('./images_api')(server.config, featureConfig)
  feature.registerApi('images', imagesApi) // exposes it on `server.features.api('images')`

  // some custom routing shown in next example
  const imageRoutes = require('./image_routes')(imagesApi, featureConfig)
  feature.registerRoutes(imageRoutes)

  done()
}
```

Regarding lifecycle plugins, I'm currently using middlewares and compose them based on some route config (see [metalman](https://www.npmjs.com/package/metalman)).
It works pretty well and there aren't almost any global middlewares (except logger/cors).

```js
module.exports = (imagesApi, config) => {
  return [{
    title: 'Get images',
    method: 'get',
    path: '/images',
    // we're just checking the scope of a jwt in here
    auth: 'images:read OR images:manage',
    // extracting common page/offset/limit query params and set it onto the request
    pagination: true,
    action: function (req, res) {
      imagesApi.getImages({
        projectId: req.verifiedAccessToken.projectId,
        limit: req.pagination.limit,
        offset: req.pagination.offset
      }, (err, images) => {
        if (err) return res.error(err)
        res.success(images)
      })
    }
  },
  {
    method: 'post',
    path: '/images',
    contentType: 'multipart/form-data',
    files: {count: 1, maxFileSize: featureConfig.uploadRestrictions.maxFileSize},
    action: function (req, res) {}
  }]
}

```

The disadvantage is that the plugin registration is still decoupled from the configuration. So you have to know the setup a bit as soon as you'd like to extend it. But it's awesome to get into it because you only need to know the configuration to create new endpoints as soon as you have a good setup.

Another approach would be to make that setup more functional:

e.g.
```js
fastify({
    method: 'get',
    path: '/images'
  },
  pagination(),
  handler: function (req, res) {}
})

fastify({
    method: 'post',
    path: '/images'
  },
  multipart({count: 1, maxFileSize: '1mb'})],
  function (req, res) {}
)
```

**Comment by delvedor** [comment]
Hi @marcbachmann, yes, we did have some discussion and we are working on the plugin implementation here #53. I just wrote some explanation, if you have some kind of doubt please let me know!

**Comment by delvedor** [comment]
Addressed in #53 and #56.
